src/recovery/mod.rs

```diff
@@ -0,0 +1,282 @@
+    fn for_step(&self, step_id: &str) -> Box<dyn StepRecovery<u64, TdPyAny, TdPyAny>>;
+
+    // TODO: We can implement state compaction / GC once we know
+    // epochs are output by every capture() operator.
+    //fn compact(&self, done_epoch: u64);
+}
+
+/// Per-operator interface for state recovery.
+///
+/// Each [`RecoveryStore`] will need to implement this with the
+/// specific queryies or inserts needed.
+pub(crate) trait StepRecovery<T, K, D> {
+    fn recover_last(&self, current_epoch: &T, key: &K) -> Option<D>;
+
+    fn save_complete(&self, completed_epoch: &T, key: &K, state: &Option<D>) -> ();
+}
+
+/// A recovery store which does nothing.
+///
+/// Saves are dropped and all recoveries result in "not found".
+pub(crate) struct NoOpRecoveryStore;
+
+impl NoOpRecoveryStore {
+    pub(crate) fn new() -> Self {
+        NoOpRecoveryStore {}
+    }
+}
+
+impl RecoveryStore for NoOpRecoveryStore {
+    fn for_step(&self, step_id: &str) -> Box<dyn StepRecovery<u64, TdPyAny, TdPyAny>> {
+        Box::new(NoOpStepRecovery {
+            step_id: step_id.to_string(),
+        })
+    }
+}
+
+struct NoOpStepRecovery {
+    step_id: String,
+}
+
+impl<T: Debug, K: Debug, D: Debug> StepRecovery<T, K, D> for NoOpStepRecovery {
+    fn recover_last(&self, current_epoch: &T, key: &K) -> Option<D> {
+        debug!(
+            "noop state fake-recovered step_id={} key={key:?}@epoch={current_epoch:?}",
+            self.step_id
+        );
+        None
+    }
+    fn save_complete(&self, completed_epoch: &T, key: &K, state: &Option<D>) -> () {
+        debug!("noop state fake-stored step_id={} key={key:?}:state={state:?}@epoch={completed_epoch:?}", self.step_id);
+        ()
+    }
+}
+
+pub(crate) struct SqliteRecoveryStore {
+    rt: SendWrapper<Rc<Runtime>>,
+    pool: Pool<Sqlite>,
+}
+
+impl SqliteRecoveryStore {
+    pub(crate) fn new(db_file_name: &str) -> Self {
+        let rt = SendWrapper::new(Rc::new(
+            tokio::runtime::Builder::new_current_thread()
+                .enable_all()
+                .build()
+                .unwrap(),
+        ));
+
+        // For some reason the busy_timeout setting doesn't affect the
+        // initial connection. So manually retry here.
+        let pool = retry(Fixed::from_millis(100), || {
+            let options = SqliteConnectOptions::new()
+                .filename(db_file_name)
+                .create_if_missing(true)
+                .busy_timeout(Duration::from_secs(5))
+                .journal_mode(sqlx::sqlite::SqliteJournalMode::Wal)
+                .locking_mode(sqlx::sqlite::SqliteLockingMode::Normal)
+                .synchronous(sqlx::sqlite::SqliteSynchronous::Normal)
+                .thread_name(|i| format!("sqlx-sqlite-{i}"));
+            let future = SqlitePoolOptions::new().connect_with(options);
+            rt.block_on(future)
+        })
+        .unwrap();
+        debug!("Opened Sqlite connection pool to {db_file_name}");
```

Comment:
I was tempted to raise a security concern here but probably a bit premature

Reply:
What specific future concerns do you have?

Reply:
i'm a nervous nelly and get 🙈  about logging any db info. It's not like it's a password tho